lib/services/role.go

```diff
@@ -714,6 +726,73 @@ type Rule struct {
 	Actions []string `json:"actions,omitempty"`
 }
 
+// CheckAndSetDefaults checks and sets defaults for this rule
+func (r *Rule) CheckAndSetDefaults() error {
+	if len(r.Resources) == 0 {
+		return trace.BadParameter("missing resources to match")
+	}
+	if len(r.Verbs) == 0 {
+		return trace.BadParameter("missing verbs")
+	}
+	if len(r.Where) != 0 {
+		parser, err := GetWhereParserFn()(&Context{})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		_, err = parser.Parse(r.Where)
+		if err != nil {
+			return trace.BadParameter("could not parse 'where' rule: %q, error: %v", r.Where, err)
+		}
+	}
+	if len(r.Actions) != 0 {
+		parser, err := GetActionsParserFn()(&Context{})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		for i, action := range r.Actions {
+			_, err = parser.Parse(action)
+			if err != nil {
+				return trace.BadParameter("could not parse action %v %q, error: %v", i, action, err)
+			}
+		}
+	}
+	return nil
+}
+
+// score is a sorting score of the rule, the more the score, the more
```

Comment:
How did you come up with the values for score below? For example why is where +8 while a wildcard is -4?

Reply:
explained in comments, take a look